### ISSUE TITLE
Wrong normalisation dimension

### ISSUE THREAD

**Issue by inspirit** [lucidrains/denoising-diffusion-pytorch#293]
Hi!

I think there is an error here:
https://github.com/lucidrains/denoising-diffusion-pytorch/blob/7a77b45870f33cc668eec4200cacb0b5df263e7b/denoising_diffusion_pytorch/karras_unet.py#L152

you are normalising across last dimension (default) which is kernel_size for convolutional weight, isn't it?

**Comment by lucidrains** [comment]
<img width="487" alt="Screen Shot 2024-02-19 at 6 22 06 AM" src="https://github.com/lucidrains/denoising-diffusion-pytorch/assets/108653/005e5618-5a2e-4338-b643-012795278905">

👋 

yes, that's how it is in Algorithm 1. if I am not mistakened?

**Comment by inspirit** [comment]
yes, it is different:
you are using 'F.normalize(t, dim = dim, eps = eps)' with default dim = -1

while in example you show above they flatten array starting dim 1 and use this last dimension to compute normalisation

**Comment by inspirit** [comment]
you are actually doing it different few lines above in training block:
https://github.com/lucidrains/denoising-diffusion-pytorch/blob/7a77b45870f33cc668eec4200cacb0b5df263e7b/denoising_diffusion_pytorch/karras_unet.py#L147

you pack weight before applying normalisation and then bring it back

**Comment by lucidrains** [comment]
@inspirit 🤦 oh yes, thank you

could you double check the latest commit?

**Comment by lucidrains** [comment]
@inspirit seems like i forgot a scaling factor too (alpha)

**Comment by inspirit** [comment]
yup looks good to me :)

**Comment by lucidrains** [comment]
@inspirit ok, do let me know how the Karras unet fairs! Tero Karras is in a league of his own, so highly optimistic on his magnitude preserving layers

**Comment by lucidrains** [comment]
@inspirit one uncertainty I had was whether the values in the cosine sim attention is also pixel normed, just putting that out there in case you have an opinion

**Comment by inspirit** [comment]
I'm adopting it to a slightly different use: generator model in adversarial learning
will see ig i can make it learn :)

**Comment by inspirit** [comment]
yup, thats exactly my thoughts

**Comment by inspirit** [comment]
i see from the paper this is intentional for gradient updates in the optimizer
